[Fidget-Spinner] iP

Fidget-Spinner opened the pull request:
# Refinement: Yet another Chatbot
## Features implemented
* Everything up till week 3.
* More entries for bullet list.

## How to get this PR running
1. Clone my [repo](https://github.com/Fidget-Spinner/ip/tree/master).
2. Open with ~VSCode~ **Intellij IDEA**
3. Build with *Gradle*.

```java
# This is a fenced code block, demonstrating my main method.
import duke.Duke;

public class Main {
    public static void main(String[] args) {
        Duke duke = new Duke();
        duke.loop();
        duke.end();
    }
}
```
## Features TODO:
- [X] Week 3
- [ ] Week 4
- [ ] Week 5
- [ ] Week 6
- [ ] Hopefully no more weeks :laughing: 

## Misc
As I once said in reaction to this Chatbot:
> Wow this Chatbot is so interesting!
> It just does the same thing everyone else's chatbot does!

Also side note `System.out.println` is really verbose :cry:

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Fidget-Spinner committed:
Level-0

Fidget-Spinner committed:
Level 1. Echo

Fidget-Spinner committed:
Level 2. Add, List

Fidget-Spinner committed:
Level 3. Mark as Done

Fidget-Spinner committed:
Level 4. ToDos, Events, Deadlines

Fidget-Spinner committed:
A-TextUiTesting: Automated Text UI Testing

Fidget-Spinner committed:
Level 5. Handle Errors

Fidget-Spinner committed:
Level 6. Delete

Fidget-Spinner committed:
Use switch-case for commands

Fidget-Spinner committed:
Level-7: Save todo list to disk

Fidget-Spinner committed:
Level 8:  Dates and Times

Fidget-Spinner committed:
Refactor to use more OOP

Fidget-Spinner committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Fidget-Spinner committed:
A-JUnit: Add JUnit tests for TaskList and Parser

Fidget-Spinner committed:
fix TaskList testRemove test

Fidget-Spinner committed:
Add support for producing jar files

Fidget-Spinner committed:
Add JavaDoc for nearly all public methods

Fidget-Spinner committed:
Move all classes to a package

Fidget-Spinner committed:
Add finding of tasks

Fidget-Spinner committed:
Merge remote-tracking branch 'origin/branch-A-CodingStandard'

Fidget-Spinner committed:
Merge remote-tracking branch 'origin/branch-Level-9'

Fidget-Spinner committed:
fix bugs from merging

Fidget-Spinner committed:
document the date parser

Fidget-Spinner committed:
reformat code

Fidget-Spinner committed:
Add CheckStyle. Fix style violations.

Fidget-Spinner committed:
Setup javaFX and add FXML files from tutorial

Fidget-Spinner committed:
Add GUI for Duke. Delete old text UI.

AdityaB4 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,42 @@
+     * @param userInput what the user entered.
+     * @return Duke's response.
+     */
+    public String getResponse(String userInput) {
+        // Inspired and referenced from
+        // https://nus-cs2103-ay2324s2.github.io/website/schedule/week3/project.html#a-moreoop
+        try {
+            Command command = Parser.parse(userInput);
+            return command.execute(this.taskList, null);
+        } catch (ChatException e) {
+            return e.getMessage();
+        } finally {
+            taskList.save();
+        }
+    }
```
Very clean way to do this!

AdityaB4 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,117 @@
+        return msg;
+    }
+
+    /**
+     * Parse a date in the format day/Month/Year HoursMinutes
+     *
+     * @param date Raw input to pass.
+     * @return A LocalDateTime object of the parse input.
+     * @throws ChatException if the input is invalid
+     */
+    private static LocalDateTime dateParse(String date) {
+        try {
+            return LocalDateTime.parse(date, DateTimeFormatter.ofPattern("d/MM/yyyy HHmm"));
+        } catch (DateTimeParseException e) {
+            throw new ChatException("Invalid date time format");
+        }
+    }
+
```
Minor nit: to comply with the coding standard, this method might be better named `parseDate` instead.

Fidget-Spinner replied:
That makes sense. Thank you.

AdityaB4 reviewed (commented):
Clean code that mostly complies with the coding standard. LGTM

Fidget-Spinner committed:
Fix build mainClass

The previous mainClass was pointing to the
template given by CS2103T. This is obviously
wrong for this project.

Fidget-Spinner committed:
Add asserts for invariants in code

The code previously assumed that certain things were
not null, and did not check for that. These assertions
make those invariants salient.

Fidget-Spinner committed:
Remove unused storage variable

Code cleanup to improve ease of understanding.

Fidget-Spinner committed:
Rename parseDate to dateParse

To comply with coding standard.

(Thanks to the review by AdityaB4)

Fidget-Spinner committed:
Merge pull request #1 from Fidget-Spinner/branch-A-CodeQuality

Code quality touchups

Fidget-Spinner committed:
Merge branch 'master' into branch-A-Assertions

Fidget-Spinner committed:
Merge pull request #2 from Fidget-Spinner/branch-A-Assertions

Assert invariants

Fidget-Spinner committed:
Add CI on all PRs and push

Includes style checks and unit tests.

Fidget-Spinner committed:
Rename Gui to gui to follow style

Fidget-Spinner committed:
Allow sorting of deadlilnes

Fidget-Spinner committed:
Make compatible with JDK 11

Fidget-Spinner committed:
Fix checkstyle violations

JuliaPoo commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,124 @@
+     */
+    public static Command parse(String line) {
+        assert line != null;
+        String command = line.split(" ")[0];
```
Might want to add length checks to handle invalid access exceptions.

JuliaPoo commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,124 @@
+        }
+        case "event": {
+            String[] res = Parser.getContents(line, 5).split(" ((/from)|(/to)) ");
+            return new EventCommand(res[0], Parser.parseDate(res[1]), Parser.parseDate(res[2]));
```
Might want to check that `/from` and `/to` appears in the correct order.

JuliaPoo commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,124 @@
+        assert line != null;
+        String msg = line.substring(startFrom);
+        if (msg.equals(" ") || msg.isEmpty()) {
+            throw new ChatException("The description of a todo cannot be empty.");
```
Outdated exception message

JuliaPoo commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+
+    @Override
+    public String toString() {
+        String isDone = this.done ? "X" : " ";
```
`isDone` is a variable name for boolean but this is a string

JuliaPoo reviewed (commented):
Very structured code. It might not be necessary to expose the different Command classes and Task classes publicly.

Fidget-Spinner committed:
Address style checks and reviews

Follow boolean naming conventions, and update docstrings to follow convention.

Fidget-Spinner committed:
Rename all ocurrences of Duke to Refinement

Align ourselves with better user branding.

Fidget-Spinner committed:
Widen user input box

Previous width was too narrow and caused text to overflow too easily

Fidget-Spinner committed:
Use smaller sized image for user

Makes the final JAR smaller.

Fidget-Spinner committed:
Add a title name for Refinement

Fidget-Spinner committed:
Increase height of main window

Fidget-Spinner committed:
Write a user guide

Fidget-Spinner committed:
Add a credits section to reference code

Fidget-Spinner committed:
Rename duke to refinement

Fidget-Spinner committed:
Fix mark/unmark docs

Fidget-Spinner committed:
Fix DialogBox min height too small

This was causing text to shrink until nothing when there were too many messages.

Fidget-Spinner committed:
Rename Duke to Refinement in README

Fidget-Spinner committed:
Make everything part of a package

Fidget-Spinner committed:
Fix mainClassName in gradle
